### ISSUE TITLE
Style/DoubleNegation: EnforcedStyle: disallow_in_conditionals

### ISSUE THREAD

**Issue by nevans** [rubocop/rubocop#10475]
## Is your feature request related to a problem? Please describe.

For similar reasons as `allowed_in_returns`, `!!foo` should be allowed in any circumstance where the distinction between `false` and `nil` (or `true` and anything else) _might_ matter.  In `rbs` types, this is the difference between `bool` and `bool?` and `boolish`.

## Describe the solution you'd like

Only disallow using `!!val` inside a conditional or a boolean expression, e.g. with `if`, `elsif`, `unless`, `when`, `until`, `?:`).

#### EnforcedStyle: disallow_in_conditionals
```ruby
# bad
foo if !!cond
foo if !!cond1 && cond2
!!cond and do_foo
val = (!!cond ? foo : bar)

if !!cond
  foo
end

while !!cond
  foo
end

# good
call(!!param)
val = !!(cond ? foo : bar)
```

## Describe alternatives you've considered

Perhaps with another config attribute, it could also be disallowed for method parameters, if we _know_ the parameter type is `boolish`.  Likewise, variables could be traced from assignment in case they are only used in a conditional context.

Or I can just add `# rubocop:disallow Style/DoubleNegation`.  Or I can replace `!!cond` with `cond ? true : false`, but that feels much worse than `!!cond` to me.

**Comment by dvandersluis** [comment]
I don't really know what use this cop would be if it wasn't checking conditionals. Sounds like for your use-case it'd be better to just disable the cop altogether?

**Comment by jonas054** [comment]
@dvandersluis I think the point @nevans makes is that `Style/DoubleNegation` is marked as `SafeAutoCorrect: false` but we could introduce an enforced style that only reports instances of `!!` where we can be sure that removing them makes no semantic difference. Thus effectively making autocorrect safe. I'm on board with that idea.

**Comment by nevans** [comment]
Yes, what @jonas054 said is exactly right.  However subtle or insignificant the distinction between `bool` and `boolish` might be, I want this cop to *only* make suggestions or corrections when the code's behavior is unchanged.  IMO, that should be the default.

I think maybe @dvandersluis read some of my examples to mean the opposite of what I intended?  Sorry if I was unclear.  It may have been unbalanced to give ten lines of disallowed cases (all of which are already disallowed by existing `EnforcedStyle`s) vs only two lines of allowed (which exhibit the change I want).

I didn't mention before, but in addition to the standard conditional clauses (`if`, `elsif`, `unless`, `case; when`, `until`, `?:`), I'd also disallow `!!val` in boolean expressions (`expr and expr`, `expr or expr`, `arg && arg`, `arg || arg`), but *only* when they are used *purely* for their short-circuiting properties.  I.e. when the boolish-vs-bool value doesn't "escape" the local context.

Maybe this is simpler to explain from a positive "only allow" perspective?  The current default style (`allow_in_returns`) is incomplete.  To keep the same semantics, we also need to allow it for method arguments and assignments.  Maybe name it `allow_arguments_assignments_and_returns` instead of `disallow_in_conditionals`? 🙂  I *think* these three cases cover every time the semantics would be changed; can you think of any others?  Maybe a better name would be something like `EnforcedStyle: allow_semantically_significant`?

For assignment to a local variable that is only ever used in conditionals, it might _seem_ like there's no difference between boolish and bool.  But even in that case there are both _communicative_ differences (to developers, and to LSP, typeprof, sorbet, etc) and _behavioral_ differences: the value is accessible from `binding` and that prevents GC!

Just like the current behavior, I would still disallow the following.  They can all be *safely* auto-corrected.
```ruby
foo if !!cond
!!cond ? foo : bar

def short_circuits_that_do_not_care_about_bool_vs_boolish
  !!cond1 and a
  !!cond2 &&  b
  !!cond3 or  c
  !!cond4 ||  d
  !!foo and
    !!bar and
    !!baz
  retval
end

stmt while !!cond
```

The following would all be allowed.  Changing them changes the semantics and is unsafe.
```ruby
foo if bar = !!baz

foo(!!bar || baz)

while okay = !!next_values
  etc
  etc
end

def short_circuits_lvar
  val = !!query1 && !!query2
  other stuff that maybe uses var
end

def short_circuits_return_val
  !!foo and
    bar and
    qux
end

def short_circuit_a = !!cond && a
def short_circuit_b = !!cond || b
def short_circuit_c;  !!cond and c end
def short_circuit_d;  !!cond or  d end

foo(!!val)
```

The following might be considered an edge case.  I don't care if the first form is allowed or disallowed... there's a case to be made either way.  My preference is whichever is simpler code and executes faster. :)
```ruby
# should these be simplified?
var1 = !!foo || !!bar || !!baz || !!quux
var2 = foo || !!bar || !!baz || quux

# semantically identical simplifications:
var1 = !!(foo || bar || baz || quux)
var2 = foo || !!(bar || baz) || quux
```
